src/block_cache/block_io.c

```diff
@@ -14,79 +14,158 @@
  */
 int
-__wt_bt_read(WT_SESSION_IMPL *session, WT_ITEM *buf, const uint8_t *addr, size_t addr_size)
+__wt_blkcache_read(WT_SESSION_IMPL *session, WT_ITEM *buf, const uint8_t *addr, size_t addr_size)
 {
+    WT_BLKCACHE *blkcache;
+    WT_BLKCACHE_ITEM *blkcache_item;
     WT_BM *bm;
     WT_BTREE *btree;
+    WT_COMPRESSOR *compressor;
     WT_DECL_ITEM(etmp);
     WT_DECL_ITEM(tmp);
     WT_DECL_RET;
     WT_ENCRYPTOR *encryptor;
     WT_ITEM *ip;
     const WT_PAGE_HEADER *dsk;
     size_t compression_ratio, result_len;
-    const char *fail_msg;
+    uint64_t time_diff, time_start, time_stop;
+    bool blkcache_found, expect_conversion, found, skip_cache, timer;
 
+    blkcache = &S2C(session)->blkcache;
+    blkcache_item = NULL;
     btree = S2BT(session);
     bm = btree->bm;
-    fail_msg = NULL; /* -Wuninitialized */
+    compressor = btree->compressor;
+    encryptor = btree->kencryptor == NULL ? NULL : btree->kencryptor->encryptor;
+    blkcache_found = found = false;
+    skip_cache = blkcache->type == BLKCACHE_UNCONFIGURED;
 
     /*
-     * If anticipating a compressed or encrypted block, read into a scratch buffer and decompress
-     * into the caller's buffer. Else, read directly into the caller's buffer.
+     * If anticipating a compressed or encrypted block, start with a scratch buffer and convert into
+     * the caller's buffer. Else, start with the caller's buffer.
      */
-    if (btree->compressor == NULL && btree->kencryptor == NULL) {
-        WT_RET(bm->read(bm, session, buf, addr, addr_size));
-        dsk = buf->data;
-        ip = NULL;
-    } else {
-        WT_RET(__wt_scr_alloc(session, 0, &tmp));
-        WT_ERR(bm->read(bm, session, tmp, addr, addr_size));
-        dsk = tmp->data;
+    ip = buf;
+    expect_conversion = compressor != NULL || encryptor != NULL;
+    if (expect_conversion) {
+        WT_RET(__wt_scr_alloc(session, 4 * 1024, &tmp));
         ip = tmp;
     }
 
-    /*
-     * If the block is encrypted, copy the skipped bytes of the original image into place, then
-     * decrypt.
-     */
-    if (F_ISSET(dsk, WT_PAGE_ENCRYPTED)) {
-        if (btree->kencryptor == NULL || (encryptor = btree->kencryptor->encryptor) == NULL ||
-          encryptor->decrypt == NULL) {
-            fail_msg = "encrypted block in file for which no encryption configured";
-            goto corrupt;
+    /* Check for mapped blocks. */
+    WT_RET(__wt_blkcache_map_read(session, ip, addr, addr_size, &found));
+    if (found) {
+        skip_cache = true;
+        if (!expect_conversion)
+            goto verify;
+    }
+
+    /* Check the block cache. */
+    if (!found && blkcache->type != BLKCACHE_UNCONFIGURED) {
+        __wt_blkcache_get(session, addr, addr_size, &blkcache_item, &found, &skip_cache);
+        if (found) {
+            blkcache_found = true;
+            ip->data = blkcache_item->data;
+            ip->size = blkcache_item->data_size;
+            if (!expect_conversion) {
```

Comment:
Hmm.  Apparently I forgot to click "save" on this comment when I finished my review.

Cached blocks are already decrypted. So if we have a decryptor but no compressor, we could be done here.

Reply:
The problem is that NVRAM cache blocks are encrypted (as you note, DRAM cache blocks are not). I went back-and-forth a couple of times changing the conditions under which I jumped to the end of the function and the conditions where I kept going through the tests. Basically, there are really only 5-6 tests (that will all fail), between this code and the end of the function, so the go-to really isn't that useful, and I think we should err strongly on the side of clarity and maintainability, and not worry about the performance of a go-to vs. running through half-a-dozen tests.

Given all that, what specific change would you propose?

Reply:
Instead of 
    `if (!expect_conversion) {`
we could say
   `if (compressor == NULL) {`

I.e., it doesn't matter encryption is configured at this point.

I'm completely on board with favoring readability over optimization, and would fine with whatever you judge best.

Reply:
I think this isn't quite right -- if the block came from an NVRAM cache, then it may need decryption -- because NVRAM persists, we can't store decrypted blocks in it, but we do store decrypted blocks in DRAM. So, later, there is:
```
    /*
     * If the block is encrypted, copy the skipped bytes of the image into place, then decrypt. DRAM
     * block-cache blocks are never encrypted.
     */
    dsk = ip->data;
    if (!blkcache_found || blkcache->type != BLKCACHE_DRAM) {
```
Or am I misunderstanding your point?

Reply:
No, I just realized I had missed that point and was typing a response when your rely arrive.